components/StateMiddleware.coffee

```diff
@@ -0,0 +1,31 @@
+noflo = require 'noflo'
+debug = require('debug') 'noflo-ui:state'
+
+exports.getComponent = ->
+  c = new noflo.Component
+  c.inPorts.add 'in',
+    datatype: 'all'
+  c.outPorts.add 'pass',
+    datatype: 'object'
+
+  c.state = {}
+  c.shutdown = ->
+    c.state = {}
+  noflo.helpers.WirePattern c,
+    in: 'in'
+    out: 'pass'
+    forwardGroups: true
+    async: true
+  , (data, groups, out, callback) ->
+    if data?.state
+      # Keep track of last state
+      c.state = data.state
+    else
+      # Warn of actions that don't contain state
+      debug "#{groups.join(':')} was sent without state, using previous state"
+    state = data?.state or c.state
```

Comment:
Explained in PR description. Basically we had problems with stale state because:

* Application state is kept in the main Polymer element
* However, certain middleware (and currently still stores) need partial state access
* For this, old stores kept their own partial state
* This state was not properly cleared in some situations, causing stores to use stale runtime connections etc
* Now latest state is sent alongside UI actions so stores/middlewares don't need to keep track of it on their own

However, in the transitional period we will have actions sent _without state_. This middleware amends them with the latest state.